File: app/src/main/java/com/micewine/emu/fragments/EditVirtualButtonFragment.java
package com.micewine.emu.fragments;

import static com.micewine.emu.activities.VirtualControllerOverlayMapper.ACTION_INVALIDATE;
import static com.micewine.emu.controller.XKeyCodes.getKeyNames;
import static com.micewine.emu.controller.XKeyCodes.getMapping;
import static com.micewine.emu.views.VirtualKeyboardInputCreatorView.ANALOG;
import static com.micewine.emu.views.VirtualKeyboardInputCreatorView.BUTTON;
import static com.micewine.emu.views.VirtualKeyboardInputCreatorView.DPAD;
import static com.micewine.emu.views.VirtualKeyboardInputCreatorView.lastSelectedButton;
import static com.micewine.emu.views.VirtualKeyboardInputCreatorView.lastSelectedType;
import static com.micewine.emu.views.VirtualKeyboardInputView.SHAPE_CIRCLE;
import static com.micewine.emu.views.VirtualKeyboardInputView.SHAPE_RECTANGLE;
import static com.micewine.emu.views.VirtualKeyboardInputView.SHAPE_SQUARE;
import static com.micewine.emu.views.VirtualKeyboardInputView.analogList;
import static com.micewine.emu.views.VirtualKeyboardInputView.buttonList;
import static com.micewine.emu.views.VirtualKeyboardInputView.dpadList;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import com.micewine.emu.R;

import java.util.List;

public class EditVirtualButtonFragment extends DialogFragment {
    @SuppressLint("SetTextI18n")
    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.fragment_edit_virtual_button, null);
        Button saveButton = view.findViewById(R.id.saveButton);
        Button cancelButton = view.findViewById(R.id.cancelButton);

        TextView radiusSeekbarValue = view.findViewById(R.id.radiusSeekbarValue);

        radiusSeekbarValue.setText(selectedButtonRadius + "%");

        SeekBar radiusSeekbar = view.findViewById(R.id.radiusSeekbar);

        radiusSeekbar.setMin(100);
        radiusSeekbar.setMax(400);
        radiusSeekbar.setProgress(selectedButtonRadius);
        radiusSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                seekBar.setProgress(seekBar.getProgress() / 5 * 5);
                radiusSeekbarValue.setText(seekBar.getProgress() + "%");
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });

        List<String> allKeyNames = getKeyNames(lastSelectedType != ANALOG);
        List<String> shapes = List.of("Circle", "Square", "Rectangle");

        Spinner shapeSpinner = view.findViewById(R.id.shapeSpinner);

        shapeSpinner.setAdapter(new ArrayAdapter<>(requireContext(), android.R.layout.simple_spinner_dropdown_item, shapes));

        int shapeId = switch (selectedButtonShape) {
            case SHAPE_CIRCLE -> 0;
            case SHAPE_SQUARE -> 1;
            case SHAPE_RECTANGLE -> 2;
            default -> -1;
        };

        shapeSpinner.setSelection(shapeId);

        Spinner buttonSpinner = view.findViewById(R.id.buttonSpinner);

        buttonSpinner.setAdapter(new ArrayAdapter<>(requireContext(), android.R.layout.simple_spinner_dropdown_item, allKeyNames));
        buttonSpinner.setSelection(allKeyNames.indexOf(selectedButtonKeyName));

        Spinner analogUpKeySpinner = view.findViewById(R.id.analogUpKeySpinner);

        analogUpKeySpinner.setAdapter(new ArrayAdapter<>(requireContext(), android.R.layout.simple_spinner_dropdown_item, allKeyNames));
        analogUpKeySpinner.setSelection(allKeyNames.indexOf(selectedAnalogUpKeyName));

        Spinner analogDownKeySpinner = view.findViewById(R.id.analogDownKeySpinner);

        analogDownKeySpinner.setAdapter(new ArrayAdapter<>(requireContext(), android.R.layout.simple_spinner_dropdown_item, allKeyNames));
        analogDownKeySpinner.setSelection(allKeyNames.indexOf(selectedAnalogDownKeyName));

        Spinner analogLeftKeySpinner = view.findViewById(R.id.analogLeftKeySpinner);

        analogLeftKeySpinner.setAdapter(new ArrayAdapter<>(requireContext(), android.R.layout.simple_spinner_dropdown_item, allKeyNames));
        analogLeftKeySpinner.setSelection(allKeyNames.indexOf(selectedAnalogLeftKeyName));

        Spinner analogRightKeySpinner = view.findViewById(R.id.analogRightKeySpinner);

        analogRightKeySpinner.setAdapter(new ArrayAdapter<>(requireContext(), android.R.layout.simple_spinner_dropdown_item, allKeyNames));
        analogRightKeySpinner.setSelection(allKeyNames.indexOf(selectedAnalogRightKeyName));

        if (lastSelectedType == ANALOG || lastSelectedType == DPAD) {
            TextView shapeText = view.findViewById(R.id.shapeText);
            TextView buttonMappingText = view.findViewById(R.id.buttonMappingText);

            shapeText.setVisibility(View.GONE);
            buttonMappingText.setVisibility(View.GONE);

            buttonSpinner.setVisibility(View.GONE);
            shapeSpinner.setVisibility(View.GONE);
        } else if (lastSelectedType == BUTTON) {
            LinearLayout layoutAnalogUp = view.findViewById(R.id.layoutAnalogUp);
            LinearLayout layoutAnalogDown = view.findViewById(R.id.layoutAnalogDown);
            LinearLayout layoutAnalogLeft = view.findViewById(R.id.layoutAnalogLeft);
            LinearLayout layoutAnalogRight = view.findViewById(R.id.layoutAnalogRight);

            layoutAnalogUp.setVisibility(View.GONE);
            layoutAnalogDown.setVisibility(View.GONE);
            layoutAnalogLeft.setVisibility(View.GONE);
            layoutAnalogRight.setVisibility(View.GONE);
        }

        saveButton.setOnClickListener((v) -> {
            if (lastSelectedType == BUTTON && !buttonList.isEmpty()) {
                buttonList.get(lastSelectedButton - 1).keyName = (buttonSpinner.getSelectedItem().toString());
                buttonList.get(lastSelectedButton - 1).buttonMapping = (getMapping(buttonSpinner.getSelectedItem().toString()));
                buttonList.get(lastSelectedButton - 1).radius = (radiusSeekbar.getProgress());

                switch (shapeSpinner.getSelectedItem().toString()) {
                    case "Circle" -> buttonList.get(lastSelectedButton - 1).shape = (SHAPE_CIRCLE);
                    case "Square" -> buttonList.get(lastSelectedButton - 1).shape = (SHAPE_SQUARE);
                    case "Rectangle" -> buttonList.get(lastSelectedButton - 1).shape = (SHAPE_RECTANGLE);
                }
            } else if (lastSelectedType == ANALOG && !analogList.isEmpty()) {
                analogList.get(lastSelectedButton - 1).upKeyName = (analogUpKeySpinner.getSelectedItem().toString());
                analogList.get(lastSelectedButton - 1).upKeyCodes = (getMapping(analogUpKeySpinner.getSelectedItem().toString()));

                analogList.get(lastSelectedButton - 1).downKeyName = (analogDownKeySpinner.getSelectedItem().toString());
                analogList.get(lastSelectedButton - 1).downKeyCodes = (getMapping(analogDownKeySpinner.getSelectedItem().toString()));

                analogList.get(lastSelectedButton - 1).leftKeyName = (analogLeftKeySpinner.getSelectedItem().toString());
                analogList.get(lastSelectedButton - 1).leftKeyCodes = (getMapping(analogLeftKeySpinner.getSelectedItem().toString()));

                analogList.get(lastSelectedButton - 1).rightKeyName = (analogRightKeySpinner.getSelectedItem().toString());
                analogList.get(lastSelectedButton - 1).rightKeyCodes = (getMapping(analogRightKeySpinner.getSelectedItem().toString()));

                analogList.get(lastSelectedButton - 1).radius = (radiusSeekbar.getProgress());
            } else if (lastSelectedType == DPAD && !dpadList.isEmpty()) {
                dpadList.get(lastSelectedButton - 1).upKeyName = (analogUpKeySpinner.getSelectedItem().toString());
                dpadList.get(lastSelectedButton - 1).upKeyCodes = (getMapping(analogUpKeySpinner.getSelectedItem().toString()));

                dpadList.get(lastSelectedButton - 1).downKeyName = (analogDownKeySpinner.getSelectedItem().toString());
                dpadList.get(lastSelectedButton - 1).downKeyCodes = (getMapping(analogDownKeySpinner.getSelectedItem().toString()));

                dpadList.get(lastSelectedButton - 1).leftKeyName = (analogLeftKeySpinner.getSelectedItem().toString());
                dpadList.get(lastSelectedButton - 1).leftKeyCodes = (getMapping(analogLeftKeySpinner.getSelectedItem().toString()));

                dpadList.get(lastSelectedButton - 1).rightKeyName = (analogRightKeySpinner.getSelectedItem().toString());
                dpadList.get(lastSelectedButton - 1).rightKeyCodes = (getMapping(analogRightKeySpinner.getSelectedItem().toString()));

                dpadList.get(lastSelectedButton - 1).radius = (radiusSeekbar.getProgress());
            }

            requireContext().sendBroadcast(new Intent(ACTION_INVALIDATE));
            dismiss();
        });

        cancelButton.setOnClickListener((v) -> dismiss());

        return new AlertDialog.Builder(requireContext(), R.style.CustomAlertDialog).setView(view).create();
    }

    public static String selectedButtonKeyName = "";
    public static String selectedAnalogUpKeyName = "";
    public static String selectedAnalogDownKeyName = "";
    public static String selectedAnalogLeftKeyName = "";
    public static String selectedAnalogRightKeyName = "";
    public static int selectedButtonShape = -1;
    public static int selectedButtonRadius = 0;
}

File: app/src/main/java/com/micewine/emu/fragments/EditVirtualButtonFragment.kt
package com.micewine.emu.fragments

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.Spinner
import android.widget.TextView
import androidx.fragment.app.DialogFragment
import com.micewine.emu.R
import com.micewine.emu.activities.VirtualControllerOverlayMapper.Companion.ACTION_INVALIDATE
import com.micewine.emu.controller.XKeyCodes.getKeyNames
import com.micewine.emu.controller.XKeyCodes.getMapping
import com.micewine.emu.views.VirtualKeyboardInputView.Companion.SHAPE_CIRCLE
import com.micewine.emu.views.VirtualKeyboardInputView.Companion.SHAPE_RECTANGLE
import com.micewine.emu.views.VirtualKeyboardInputView.Companion.SHAPE_SQUARE
import com.micewine.emu.views.VirtualKeyboardInputView.Companion.analogList
import com.micewine.emu.views.VirtualKeyboardInputView.Companion.buttonList
import com.micewine.emu.views.VirtualKeyboardInputView.Companion.dpadList
import com.micewine.emu.views.VirtualKeyboardInputCreatorView.Companion.ANALOG
import com.micewine.emu.views.VirtualKeyboardInputCreatorView.Companion.BUTTON
import com.micewine.emu.views.VirtualKeyboardInputCreatorView.Companion.DPAD
import com.micewine.emu.views.VirtualKeyboardInputCreatorView.Companion.lastSelectedButton
import com.micewine.emu.views.VirtualKeyboardInputCreatorView.Companion.lastSelectedType

class EditVirtualButtonFragment : DialogFragment() {
    @SuppressLint("SetTextI18n")
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val inflater = requireActivity().layoutInflater
        val view = inflater.inflate(R.layout.fragment_edit_virtual_button, null)
        val saveButton = view.findViewById<Button>(R.id.saveButton)
        val cancelButton = view.findViewById<Button>(R.id.cancelButton)
        val radiusSeekbarValue = view.findViewById<TextView>(R.id.radiusSeekbarValue).apply {
            text = "$selectedButtonRadius%"
        }
        val radiusSeekbar = view.findViewById<SeekBar>(R.id.radiusSeekbar).apply {
            max = 400
            min = 100

            setOnSeekBarChangeListener(object :
                OnSeekBarChangeListener {
                @SuppressLint("SetTextI18n")
                override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                    seekBar?.progress = Math.round(progress.toFloat() / 5) * 5
                    radiusSeekbarValue?.text = "${seekBar?.progress}%"
                }

                override fun onStartTrackingTouch(seekBar: SeekBar?) {
                }

                override fun onStopTrackingTouch(seekBar: SeekBar?) {
                }
            })
        }

        val allKeyNames = getKeyNames(lastSelectedType != ANALOG)
        val shapes = listOf("Circle", "Square", "Rectangle")

        val shapeSpinner = view.findViewById<Spinner>(R.id.shapeSpinner).apply {
            adapter = ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, shapes)
            setSelection(
                when (selectedButtonShape) {
                    SHAPE_CIRCLE -> 0
                    SHAPE_SQUARE -> 1
                    SHAPE_RECTANGLE -> 2
                    else -> -1
                }
            )
        }
        val buttonSpinner = view.findViewById<Spinner>(R.id.buttonSpinner).apply {
            adapter = ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, allKeyNames)
            setSelection(allKeyNames.indexOf(selectedButtonKeyName))
        }
        val analogUpKeySpinner = view.findViewById<Spinner>(R.id.analogUpKeySpinner).apply {
            adapter = ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, allKeyNames)
            setSelection(allKeyNames.indexOf(selectedAnalogUpKeyName))
        }
        val analogDownKeySpinner = view.findViewById<Spinner>(R.id.analogDownKeySpinner).apply {
            adapter = ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, allKeyNames)
            setSelection(allKeyNames.indexOf(selectedAnalogDownKeyName))
        }
        val analogLeftKeySpinner = view.findViewById<Spinner>(R.id.analogLeftKeySpinner).apply {
            adapter = ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, allKeyNames)
            setSelection(allKeyNames.indexOf(selectedAnalogLeftKeyName))
        }
        val analogRightKeySpinner = view.findViewById<Spinner>(R.id.analogRightKeySpinner).apply {
            adapter = ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, allKeyNames)
            setSelection(allKeyNames.indexOf(selectedAnalogRightKeyName))
        }

        radiusSeekbar.progress = selectedButtonRadius

        if (lastSelectedType == ANALOG || lastSelectedType == DPAD) {
            view.findViewById<TextView>(R.id.shapeText).visibility = View.GONE
            view.findViewById<TextView>(R.id.buttonMappingText).visibility = View.GONE

            buttonSpinner.visibility = View.GONE
            shapeSpinner.visibility = View.GONE
        } else if (lastSelectedType == BUTTON) {
            view.findViewById<LinearLayout>(R.id.layoutAnalogUp).visibility = View.GONE
            view.findViewById<LinearLayout>(R.id.layoutAnalogDown).visibility = View.GONE
            view.findViewById<LinearLayout>(R.id.layoutAnalogLeft).visibility = View.GONE
            view.findViewById<LinearLayout>(R.id.layoutAnalogRight).visibility = View.GONE
        }

        val dialog = AlertDialog.Builder(requireContext(), R.style.CustomAlertDialog).setView(view).create()

        saveButton.setOnClickListener {
            if (lastSelectedType == BUTTON && buttonList.isNotEmpty()) {
                buttonList[lastSelectedButton - 1].keyName = buttonSpinner.selectedItem.toString()
                buttonList[lastSelectedButton - 1].buttonMapping = getMapping(buttonSpinner.selectedItem.toString())

                buttonList[lastSelectedButton - 1].radius = radiusSeekbar.progress.toFloat()

                when (shapeSpinner.selectedItem.toString()) {
                    "Circle" -> buttonList[lastSelectedButton - 1].shape = SHAPE_CIRCLE
                    "Square" -> buttonList[lastSelectedButton - 1].shape = SHAPE_SQUARE
                    "Rectangle" -> buttonList[lastSelectedButton - 1].shape = SHAPE_RECTANGLE
                }
            } else if (lastSelectedType == ANALOG && analogList.isNotEmpty()) {
                analogList[lastSelectedButton - 1].upKeyName = analogUpKeySpinner.selectedItem.toString()
                analogList[lastSelectedButton - 1].upKeyCodes = getMapping(analogUpKeySpinner.selectedItem.toString())

                analogList[lastSelectedButton - 1].downKeyName = analogDownKeySpinner.selectedItem.toString()
                analogList[lastSelectedButton - 1].downKeyCodes = getMapping(analogDownKeySpinner.selectedItem.toString())

                analogList[lastSelectedButton - 1].leftKeyName = analogLeftKeySpinner.selectedItem.toString()
                analogList[lastSelectedButton - 1].leftKeyCodes = getMapping(analogLeftKeySpinner.selectedItem.toString())

                analogList[lastSelectedButton - 1].rightKeyName = analogRightKeySpinner.selectedItem.toString()
                analogList[lastSelectedButton - 1].rightKeyCodes = getMapping(analogRightKeySpinner.selectedItem.toString())

                analogList[lastSelectedButton - 1].radius = radiusSeekbar.progress.toFloat()
            } else if (lastSelectedType == DPAD && dpadList.isNotEmpty()) {
                dpadList[lastSelectedButton - 1].upKeyName = analogUpKeySpinner.selectedItem.toString()
                dpadList[lastSelectedButton - 1].upKeyCodes = getMapping(analogUpKeySpinner.selectedItem.toString())

                dpadList[lastSelectedButton - 1].downKeyName = analogDownKeySpinner.selectedItem.toString()
                dpadList[lastSelectedButton - 1].downKeyCodes = getMapping(analogDownKeySpinner.selectedItem.toString())

                dpadList[lastSelectedButton - 1].leftKeyName = analogLeftKeySpinner.selectedItem.toString()
                dpadList[lastSelectedButton - 1].leftKeyCodes = getMapping(analogLeftKeySpinner.selectedItem.toString())

                dpadList[lastSelectedButton - 1].rightKeyName = analogRightKeySpinner.selectedItem.toString()
                dpadList[lastSelectedButton - 1].rightKeyCodes = getMapping(analogRightKeySpinner.selectedItem.toString())

                dpadList[lastSelectedButton - 1].radius = radiusSeekbar.progress.toFloat()
            }

            context?.sendBroadcast(
                Intent(ACTION_INVALIDATE)
            )

            dismiss()
        }

        cancelButton.setOnClickListener {
            dismiss()
        }

        return dialog
    }

    companion object {
        var selectedButtonKeyName = ""
        var selectedAnalogUpKeyName = ""
        var selectedAnalogDownKeyName = ""
        var selectedAnalogLeftKeyName = ""
        var selectedAnalogRightKeyName = ""
        var selectedButtonShape = -1
        var selectedButtonRadius = 0
    }
}